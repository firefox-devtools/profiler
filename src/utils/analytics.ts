/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * Document Google Analytics API that is used in the project. These definitions
 * can be updated as more features are used.
 *
 * https://developers.google.com/analytics/devguides/collection/analyticsjs/pages
 */
type GAEvent = {
  hitType: 'event',
  // Specifies the event category. Must not be empty
  eventCategory: string,
  eventAction: string,
  eventLabel?: string,
  eventValue?: number,
};

type GAPageView = {
  hitType: 'pageview',
  page: string,
};

type GATiming = {
  hitType: 'timing',
  timingCategory: string,
  timingVar: string,
  timingValue: number,
  timingLabel?: string,
};

export type GAPayload = GAEvent | GAPageView | GATiming;

export type GAErrorPayload = {
  readonly exDescription: string,
  readonly exFatal: boolean,
};

// Prettier breaks with multiple arrow functions and intersections, so name the arrow
// functions.
type _Send = (command: 'send', payload: GAPayload) => void;
type _Exception = (command: 'send', type: 'exception', payload: GAErrorPayload) => void;
export type GoogleAnalytics = _Send & _Exception;

export function sendAnalytics(payload: GAPayload) {
  const ga: GoogleAnalytics | null = (self as any).ga;
  if (ga) {
    ga('send', payload);
  }
}

export function reportError(errorPayload: GAErrorPayload) {
  const ga: GoogleAnalytics | null = (self as any).ga;
  if (ga) {
    ga('send', 'exception', errorPayload);
  }
}