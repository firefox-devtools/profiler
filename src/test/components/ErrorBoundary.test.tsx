/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { stripIndent } from 'common-tags';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import {
  NonLocalizedErrorBoundary,
  LocalizedErrorBoundary,
} from '../../components/app/ErrorBoundary';
import { withAnalyticsMock } from '../fixtures/mocks/analytics';

describe('app/NonLocalizedErrorBoundary', function () {
  const childComponentText = 'This is a child component';
  const friendlyErrorMessage = 'Oops, there was an error';
  const technicalErrorMessage = 'This is an error.';
  const ThrowingComponent = () => {
    throw new Error(technicalErrorMessage);
  };

  function setupComponent(childComponent: React.ReactNode) {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {});
    const results = render(
      <NonLocalizedErrorBoundary message={friendlyErrorMessage}>
        {childComponent}
      </NonLocalizedErrorBoundary>
    );
    return { spy, ...results };
  }

  it('matches the snapshot', () => {
    const { container } = setupComponent(<ThrowingComponent />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('shows the component children when there is no error', () => {
    const { getByText } = setupComponent(childComponentText);
    expect(getByText(childComponentText)).toBeInTheDocument();
  });

  it('shows the error message children when the component throws error', () => {
    const { getByText } = setupComponent(<ThrowingComponent />);
    expect(getByText(new RegExp(friendlyErrorMessage))).toBeInTheDocument();
  });

  it('surfaces the error via console.error', () => {
    setupComponent(<ThrowingComponent />);
    expect(console.error).toHaveBeenCalled();
  });

  it('reports errors to the analytics', () => {
    withAnalyticsMock(() => {
      setupComponent(<ThrowingComponent />);
      expect(self.ga).toHaveBeenCalledWith('send', 'exception', {
        exDescription: expect.stringMatching(
          // This regexp looks a  bit scary, but all it does is avoiding
          // matching on things that depends on the environment (like path
          // names) and path separators (unixes use `/` but windows uses `\`).
          new RegExp(stripIndent`
              Error: This is an error\\.

                  at ThrowingComponent \\(.*[/\\\\]ErrorBoundary.test.tsx:.*\\)
                  at ErrorBoundaryInternal \\(.*[/\\\\]ErrorBoundary.tsx:.*\\)
                  at NonLocalizedErrorBoundary
                  at LocalizationProvider \\(.*[/\\\\]@fluent[/\\\\]react[/\\\\]index.js:.*\\)
          `)
        ),
        exFatal: true,
      });
    });
  });
});

describe('app/LocalizedErrorBoundary', function () {
  const friendlyErrorMessage = 'Oops, there was an error';
  const technicalErrorMessage = 'This is an error.';
  const ThrowingComponent = () => {
    throw new Error(technicalErrorMessage);
  };

  function setupComponent(childComponent: React.ReactNode) {
    const spy = jest.spyOn(console, 'error').mockImplementation(() => {});
    const results = render(
      <LocalizedErrorBoundary message={friendlyErrorMessage}>
        {childComponent}
      </LocalizedErrorBoundary>
    );
    return { spy, ...results };
  }

  it('matches the snapshot', () => {
    setupComponent(<ThrowingComponent />);
    expect(document.body).toMatchSnapshot();
  });
});
