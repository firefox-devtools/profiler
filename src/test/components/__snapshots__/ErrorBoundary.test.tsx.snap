// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`app/LocalizedErrorBoundary matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="appErrorBoundary"
    >
      <div
        class="appErrorBoundaryContents"
      >
        <div
          class="photon-message-bar photon-message-bar-error photon-message-bar-inner-content"
        >
          <div
            class="photon-message-bar-inner-text appErrorBoundaryInnerText"
          >
            <p>
              Oops, there was an error
            </p>
            <p>
              Error: This is an error.
              .
            </p>
            <p>
              Please report this issue to the developers, including the full
error as displayed in the Developer Tools’ Web Console.
            </p>
          </div>
          <a
            class="photon-button photon-button-micro photon-message-bar-action-button"
            href="https://github.com/firefox-devtools/profiler/issues/new?title=An%20error%20occurred%20in%20Firefox%20Profiler"
            rel="noreferrer"
            target="_blank"
          >
            Report the error on GitHub
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`app/NonLocalizedErrorBoundary matches the snapshot 1`] = `
<div
  class="appErrorBoundary"
>
  <div
    class="appErrorBoundaryContents"
  >
    <div
      class="photon-message-bar photon-message-bar-error photon-message-bar-inner-content"
    >
      <div
        class="photon-message-bar-inner-text appErrorBoundaryInnerText"
      >
        <p>
          Oops, there was an error
        </p>
        <p>
          Error: This is an error.
          .
        </p>
        <p>
          Please report this error to the developers, including the full error as displayed in the Developer Tools’ Web Console.
        </p>
      </div>
      <a
        class="photon-button photon-button-micro photon-message-bar-action-button"
        href="https://github.com/firefox-devtools/profiler/issues/new?title=An%20error%20occurred%20in%20Firefox%20Profiler"
        rel="noreferrer"
        target="_blank"
      >
        Report the error on GitHub
      </a>
    </div>
  </div>
</div>
`;
